### PR TITLE
Implement RFC 2396: `#[target_feature]` 1.1

### PR DESCRIPTION
**Pull request by LeSeulArtichaut**
Tracking issue: #69098

r? @nikomatsakis
cc @gnzlbg @joshtriplett

### CODE REVIEW

File: src/librustc/ty/error.rs [hunk 1]
```diff
@@ -58,6 +58,8 @@ pub enum TypeError<'tcx> {
     ConstMismatch(ExpectedFound<&'tcx ty::Const<'tcx>>),
 
     IntrinsicCast,
+    /// Safe `#[target_feature]` functions are not assignable to safe function pointers.
+    TargetFeatureCast,
```
Comment (LeSeulArtichaut) [inline src/librustc/ty/error.rs]: Not sure if it belongs here, or should be part of `TypeError::Mismatch`

File: src/librustc_mir/transform/check_unsafety.rs [hunk 2]
```diff
@@ -20,6 +20,7 @@ use crate::util;
 
 pub struct UnsafetyChecker<'a, 'tcx> {
     body: &'a Body<'tcx>,
+    body_did: DefId,
```
Comment (LeSeulArtichaut) [inline src/librustc_mir/transform/check_unsafety.rs]: Should I move this to the `SourceInfo`, or is it fine here?

**Comment by LeSeulArtichaut** [comment]
@nikomatsakis being busy, I don't know who to assign.
r? @Centril can you assign someone for me, or review yourself? Thank you 😄

File: src/test/ui/rfcs/rfc-2396-target_feature-fn-ptr.rs [hunk 3]
```diff
@@ -0,0 +1,9 @@
+#[target_feature(enable = "sse2")]
```
Comment (Centril) [inline src/test/ui/rfcs/rfc-2396-target_feature-fn-ptr.rs]: Please move all relevant tests for the feature to `src/test/ui/rfc-2396-target_feature/`.

File: src/librustc_typeck/check/coercion.rs [hunk 4]
```diff
@@ -692,12 +692,22 @@ impl<'f, 'tcx> Coerce<'f, 'tcx> {
...
         match b.kind {
-            ty::FnPtr(_) => {
+            ty::FnPtr(b_sig) => {
                 let a_sig = a.fn_sig(self.tcx);
                 // Intrinsics are not coercible to function pointers
                 if a_sig.abi() == Abi::RustIntrinsic || a_sig.abi() == Abi::PlatformIntrinsic {
                     return Err(TypeError::IntrinsicCast);
                 }
+
+                // Safe #[target_feature] functions are not assignable to safe fn pointers (RFC 2396)
```
Comment (Centril) [inline src/librustc_typeck/check/coercion.rs]:
```suggestion
                // Safe `#[target_feature]` functions are not assignable to safe fn pointers (RFC 2396).
```

File: src/test/ui/rfcs/rfc-2396-target_feature-safe-calls.rs [hunk 5]
```diff
@@ -0,0 +1,19 @@
+#[target_feature(enable = "sse2")]
```
Comment (Centril) [inline src/test/ui/rfcs/rfc-2396-target_feature-safe-calls.rs]: Seems like this isn't feature gated. Let's fix that and add a feature gate test in another file.

File: src/test/ui/rfcs/rfc-2396-target_feature-safe-calls.rs [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+fn foo() {}
+
+#[target_feature(enable = "sse2")]
+fn bar() {
+    foo();
+}
+
+#[target_feature(enable = "avx")]
```
Comment (Centril) [inline src/test/ui/rfcs/rfc-2396-target_feature-safe-calls.rs]: Let's also test a case where there's no attribute.

Comment (Centril) [inline src/test/ui/rfcs/rfc-2396-target_feature-safe-calls.rs]: Let's also do a case of more than just one where the caller has one, missing one.

File: src/test/rustdoc/doc-cfg-target-feature.rs [hunk 7]
```diff
@@ -18,4 +18,4 @@
 /// assert!(false);
 /// ```
 #[doc(cfg(target_feature = "sse"))]
-pub unsafe fn foo() {}
+pub fn foo() {}
```
Comment (Centril) [inline src/test/rustdoc/doc-cfg-target-feature.rs]: Can we extend the test rather than replacing?

File: src/test/rustdoc/doc-cfg.rs [hunk 8]
```diff
@@ -48,7 +48,7 @@ pub mod unix_only {
 // @has - '//*[@id="main"]/*[@class="stability"]/*[@class="stab portability"]' \
 //        'This is supported with target feature avx only.'
 #[target_feature(enable = "avx")]
-pub unsafe fn uses_target_feature() {
+pub fn uses_target_feature() {
```
Comment (Centril) [inline src/test/rustdoc/doc-cfg.rs]: Same, let's extend?

File: src/test/ui/rfcs/rfc-2396-target_feature-safe-calls.rs (hunk 5 above)
Comment (Centril) [inline src/test/ui/rfcs/rfc-2396-target_feature-safe-calls.rs]: There doesn't seem to be any run-pass tests here; please add some. In particular, it should exercise:
- Exact matches (needs / provides is the exact same set), including with one feature and with say 3 features.
- Subset matches (needs N, provides N+1, for e.g. N = 1, 2).

File: src/test/ui/rfcs/rfc-2396-target_feature-safe-calls.rs [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+fn baz() {
+    foo(); //~ ERROR call to function with #[target_feature] is unsafe
+}
+
+fn main() {
+    unsafe {
```
Comment (Centril) [inline src/test/ui/rfcs/rfc-2396-target_feature-safe-calls.rs]: Let's also test:
> Calling them from other contexts (other functions, static variable initializers, etc.) requires opening an unsafe {} even though they are not marked as unsafe

in particular with respect to various const contexts (e.g. static and const initializers).

File: src/test/ui/rfcs/rfc-2396-target_feature-safe-calls.rs [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+        bar();
+        baz();
+    }
+}
```
Comment (Centril) [inline src/test/ui/rfcs/rfc-2396-target_feature-safe-calls.rs]: In a separate test, let's also exercise `#[target_feature]` on functions in trait definitions and in trait impls.

File: src/test/ui/rfcs/rfc-2396-target_feature-fn-ptr.rs [hunk 11]
```diff
@@ -0,0 +1,9 @@
...
+fn foo() {}
```
Comment (Centril) [inline src/test/ui/rfcs/rfc-2396-target_feature-fn-ptr.rs]: Let's also exercise this for inherent functions.

File: src/test/ui/rfcs/rfc-2396-target_feature-safe-calls.rs [hunk 12]
```diff
@@ -0,0 +1,19 @@
+#[target_feature(enable = "sse2")]
+fn foo() {}
+
+#[target_feature(enable = "sse2")]
```
Comment (Centril) [inline src/test/ui/rfcs/rfc-2396-target_feature-safe-calls.rs]: And this too, let's exercise it for inherent methods (including in the run-pass test).

File: src/librustc_mir/transform/check_unsafety.rs [hunk 13]
```diff
@@ -88,6 +91,21 @@ impl<'a, 'tcx> Visitor<'tcx> for UnsafetyChecker<'a, 'tcx> {
...
+                            self.tcx.codegen_fn_attrs(self.body_did).target_features;
+                        if !callee_features.iter().all(|feature| self_features.contains(feature)) {
+                            self.require_unsafe(
+                                "call to function with #[target_feature]",
+                                "make sure that this can only be called if the required features are available",
+                                UnsafetyViolationKind::General,
+                            )
+                        }
+                    }
+                }
```
Comment (Centril) [inline src/librustc_mir/transform/check_unsafety.rs]: Let's extract this to its own method.

File: src/librustc_mir/transform/check_unsafety.rs [hunk 14]
```diff
@@ -88,6 +91,21 @@ impl<'a, 'tcx> Visitor<'tcx> for UnsafetyChecker<'a, 'tcx> {
                         UnsafetyViolationKind::GeneralAndConstFn,
                     )
                 }
+
+                if let ty::FnDef(func_id, _) = func_ty.kind {
+                    let callee_features = self.tcx.codegen_fn_attrs(func_id).target_features;
+                    if !callee_features.is_empty() {
+                        let self_features =
+                            self.tcx.codegen_fn_attrs(self.body_did).target_features;
```
Comment (Centril) [inline src/librustc_mir/transform/check_unsafety.rs]: Consider adding a method like `x_codgen_attrs.is_target_features_subset(y_codegen_attrs)` to express the intent in a less low level fashion. (Or if you don't think that will be reused, leave a comment here.)

File: src/librustc_mir/transform/check_unsafety.rs [hunk 15]
```diff
@@ -88,6 +91,21 @@ impl<'a, 'tcx> Visitor<'tcx> for UnsafetyChecker<'a, 'tcx> {
...
+                        if !callee_features.iter().all(|feature| self_features.contains(feature)) {
+                            self.require_unsafe(
+                                "call to function with #[target_feature]",
```
Comment (Centril) [inline src/librustc_mir/transform/check_unsafety.rs]:
```suggestion
                                "call to function with `#[target_feature]`",
```

File: src/librustc_mir/transform/check_unsafety.rs [hunk 16]
```diff
@@ -88,6 +91,21 @@ impl<'a, 'tcx> Visitor<'tcx> for UnsafetyChecker<'a, 'tcx> {
...
+                                "make sure that this can only be called if the required features are available",
```
Comment (Centril) [inline src/librustc_mir/transform/check_unsafety.rs]:
```suggestion
                                "can only be called if the required target features are available",
```

File: src/librustc_typeck/check/coercion.rs [hunk 17]
```diff
@@ -692,12 +692,22 @@ impl<'f, 'tcx> Coerce<'f, 'tcx> {
...
+                if let ty::FnDef(def_id, _) = a.kind {
+                    if b_sig.unsafety() == hir::Unsafety::Normal
+                        && !self.tcx.codegen_fn_attrs(def_id).target_features.is_empty()
+                    {
+                        return Err(TypeError::TargetFeatureCast);
```
Comment (Centril) [inline src/librustc_typeck/check/coercion.rs]: Diagnostics nit: saving the spans to the `target_feature` attributes would make for good diagnostics here.

File: src/librustc/ty/error.rs [hunk 18]
```diff
@@ -489,6 +497,11 @@ impl Trait for X {
                     );
                 }
             }
+            TargetFeatureCast => {
+                db.note(
+                    "functions with `#[target_feature]` can only be assigned to `unsafe` function pointers"
+                );
```
Comment (Centril) [inline src/librustc/ty/error.rs]:
```suggestion
                );
                db.span_labels(target_spans, "`#[target_feature]` added here");
```

File: src/librustc/ty/error.rs [hunk 19]
```diff
@@ -489,6 +497,11 @@ impl Trait for X {
                     );
                 }
             }
+            TargetFeatureCast => {
```
Comment (Centril) [inline src/librustc/ty/error.rs]:
```suggestion
            TargetFeatureCast(target_spans) => {
```

File: src/librustc_typeck/collect.rs [hunk 20]
```diff
@@ -2301,14 +2301,6 @@ fn codegen_fn_attrs(tcx: TyCtxt<'_>, id: DefId) -> CodegenFnAttrs {
                 codegen_fn_attrs.export_name = Some(s);
             }
         } else if attr.check_name(sym::target_feature) {
-            if tcx.is_closure(id) || tcx.fn_sig(id).unsafety() == Unsafety::Normal {
-                let msg = "`#[target_feature(..)]` can only be applied to `unsafe` functions";
-                tcx.sess
-                    .struct_span_err(attr.span, msg)
-                    .span_label(attr.span, "can only be applied to `unsafe` functions")
-                    .span_label(tcx.def_span(id), "not an `unsafe` function")
-                    .emit();
```
Comment (Centril) [inline src/librustc_typeck/collect.rs]: Here would be the place to do the feature gating with `feature_err` (https://doc.rust-lang.org/nightly/nightly-rustc/rustc_session/parse/fn.feature_err.html).

File: src/librustc_typeck/check/coercion.rs (hunk 17 above)
Comment (LeSeulArtichaut) [reply src/librustc_typeck/check/coercion.rs]: How do I get the spans? Do I need to modify the `codegen_fn_attrs` query?

Comment (Centril) [reply src/librustc_typeck/check/coercion.rs]: Modifying the query might cause perf regressions in the happy path (but maybe worth testing). Otherwise, you have the `def_id`, so you can get the attributes via `tcx.get_attrs(def_id)` and then filter those attributes for `sym::target_feature` and extract the spans of the attributes. It should be too complicated, but it would be good to extract this logic, including `if let ty::FnDef(...)`, to a function `check_target_feature_to_fnptr`.

File: src/librustc_feature/active.rs [hunk 21]
```diff
@@ -552,6 +552,9 @@ declare_features! (
     /// Allows the use of `no_sanitize` attribute.
     (active, no_sanitize, "1.42.0", Some(39699), None),
 
+    /// Allows the use of `#[target_feature]` on safe functions
```
Comment (Centril) [inline src/librustc_feature/active.rs]:
```suggestion
    /// Allows the use of `#[target_feature]` on safe functions.
```

**Comment by LeSeulArtichaut** [comment]
Uh, found a corner case. Consider this:

```Rust
#![feature(target_feature_11)]
trait Foo {
    fn sse2(&self);
}

struct Bar;

impl Foo for Bar {
    #[target_feature(enable = "sse2")]
    fn sse2(&self) {}
}

fn foo<T: Foo>(t: &T) {
    <T as Foo>::sse2(t);
}

fn main() {
    foo(&Bar); // no `unsafe`, and no guarantee `sse2` is available
}
```

Under my current implementation, this will end up calling `sse2`, going through the `unsafe` limitation.
@gnzlbg @Centril what do you think? How should I handle this? Should I forbid `#[target_feature]` on trait implementations? Or should I do somethink after monomorphizing?

File: src/librustc/ty/error.rs [hunk 22]
```diff
@@ -489,6 +497,12 @@ impl Trait for X {
                     );
                 }
             }
+            TargetFeatureCast(ref target_spans) => {
+                db.note(
+                    "functions with `#[target_feature]` can only be assigned to `unsafe` function pointers"
+                );
+                db.span_labels(target_spans.iter().cloned(), "`#[target_feature]` added here");
```
Comment (Centril) [inline src/librustc/ty/error.rs]:
```suggestion
                db.span_labels(target_spans.iter().copied(), "`#[target_feature]` added here");
```

File: src/librustc/ty/error.rs [hunk 23]
```diff
@@ -58,6 +58,8 @@ pub enum TypeError<'tcx> {
     ConstMismatch(ExpectedFound<&'tcx ty::Const<'tcx>>),
 
     IntrinsicCast,
+    /// Safe `#[target_feature]` functions are not assignable to safe function pointers.
+    TargetFeatureCast(Vec<Span>),
```
Comment (Centril) [inline src/librustc/ty/error.rs]: Hmm, maybe store a `Box<[Span]>` instead as we're never mutating the collection.

File: src/test/ui/rfcs/rfc-2396-target_feature-11/fn-ptr.rs [hunk 24]
```diff
@@ -0,0 +1,11 @@
+#![feature(target_feature_11)]
+
+#[target_feature(enable = "sse2")]
+fn foo() {}
+
+fn main() {
+    let foo_unsf: unsafe fn() = foo;
```
Comment (Centril) [inline src/test/ui/rfcs/rfc-2396-target_feature-11/fn-ptr.rs]: Can you move this to a separate `check-pass` test?

File: src/test/ui/target-feature/invalid-attribute.stderr [hunk 25]
```diff
@@ -22,14 +22,17 @@ error: malformed `target_feature` attribute input
...
    |
 LL | #[target_feature(enable = "sse2")]
-   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ can only be applied to `unsafe` functions
+   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 ...
 LL | fn bar() {}
    | ----------- not an `unsafe` function
+   |
+   = note: see issue #69098 <https://github.com/rust-lang/rust/issues/69098> for more information
+   = help: add `#![feature(target_feature_11)]` to the crate attributes to enable
```
Comment (Centril) [inline src/test/ui/target-feature/invalid-attribute.stderr]: This can be removed from this test now to have a canonical place for the gating testing.

File: src/test/ui/rfcs/rfc-2396-target_feature-11/safe-calls.rs [hunk 26]
```diff
@@ -0,0 +1,80 @@
...
+struct Quux;
+
+impl Quux {
+    #[target_feature(enable = "avx")]
+    #[target_feature(enable = "bmi2")]
+    fn avx_bmi2(&self) {}
+}
+
+impl FooBar for Quux {
+    #[target_feature(enable = "sse2")]
```
Comment (Centril) [inline src/test/ui/rfcs/rfc-2396-target_feature-11/safe-calls.rs]: According to the RFC this should be an error.

File: src/test/ui/rfcs/rfc-2396-target_feature-11/safe-calls.rs [hunk 27]
```diff
@@ -0,0 +1,80 @@
+#![feature(target_feature_11)]
```
Comment (Centril) [inline src/test/ui/rfcs/rfc-2396-target_feature-11/safe-calls.rs]: I would like to see a separate run-pass test for the parts that are not supposed to error.

File: src/librustc/ty/error.rs (hunk 1 above)
Comment (Centril) [reply src/librustc/ty/error.rs]: It seems fine to me like this.

File: src/test/ui/rfcs/rfc-2396-target_feature-11/safe-calls.rs (hunk 26 above)
Comment (LeSeulArtichaut) [reply src/test/ui/rfcs/rfc-2396-target_feature-11/safe-calls.rs]: Oops! I missed this part of the RFC...

File: src/librustc_typeck/collect.rs [hunk 28]
```diff
@@ -2339,13 +2339,36 @@ fn codegen_fn_attrs(tcx: TyCtxt<'_>, id: DefId) -> CodegenFnAttrs {
...
+                                .struct_span_err(
+                                    attr.span,
+                                    "`#[target_feature(..)]` cannot be applied to safe trait method",
+                                )
+                                .span_label(attr.span, "cannot be applied to safe trait method")
+                                .span_label(tcx.def_span(id), "not an `unsafe` function")
+                                .emit();
+                        }
+                    }
+                }
```
Comment (LeSeulArtichaut) [inline src/librustc_typeck/collect.rs]: I will do some refractoring here

**Comment by LeSeulArtichaut** [comment]
Running the compiler directly gives me the expected error
```
error: `#[target_feature(..)]` cannot be applied to safe trait method
```

But running using `x.py` makes the code compile with no error. What did I miss? 😅

**Comment by JohnCSimon** [comment]
Ping from triage: 
@LeSeulArtichaut any updates on this?

**Comment by LeSeulArtichaut** [comment]
I'm still stuck on https://github.com/rust-lang/rust/pull/69274#issuecomment-589970615 :(

File: src/librustc/ty/error.rs (hunk 23 above)
Comment (LeSeulArtichaut) [reply src/librustc/ty/error.rs]: The `TargetFeatureCast` is created here:

https://github.com/rust-lang/rust/blob/553e7ab7d997f72588c107c2346934c373c55005/src/librustc_typeck/check/coercion.rs#L706-L713

So I'm not sure if storing a slice is a good idea, nor do I know precisely how to do it.

Comment (Centril) [reply src/librustc/ty/error.rs]: Given `impl<A> FromIterator<A> for Box<[A]>` you don't need to change that code; just change the stored type and it should just work.

Comment (LeSeulArtichaut) [reply src/librustc/ty/error.rs]: I didn't know `Box` implemented `FromIterator` :eyes:. Makes a lot more sense now 😅

File: src/librustc_feature/active.rs [hunk 29]
```diff
@@ -555,6 +555,9 @@ declare_features! (
     // Allows limiting the evaluation steps of const expressions
     (active, const_eval_limit, "1.43.0", Some(67217), None),
 
+    /// Allows the use of `#[target_feature]` on safe functions.
+    (active, target_feature_11, "1.43.0", Some(69098), None),
```
Comment (LeSeulArtichaut) [inline src/librustc_feature/active.rs]: Note to myself: change this to 1.44.0
```suggestion
    (active, target_feature_11, "1.44.0", Some(69098), None),
```

**Review by hanna-kruppe — CHANGES_REQUESTED** [review]

File: src/librustc_mir/transform/check_unsafety.rs [hunk 30]
```diff
@@ -439,6 +446,24 @@ impl<'a, 'tcx> UnsafetyChecker<'a, 'tcx> {
             }
         }
     }
+
+    /// Checks whether calling `func_did` needs an `unsafe` context or not, i.e. whether
+    /// the called function has target features the calling function hasn't.
+    fn check_target_features(&mut self, func_did: DefId) {
+        let callee_features = self.tcx.codegen_fn_attrs(func_did).target_features;
+        if !callee_features.is_empty() {
```
[the remark this answers is not in the document]
Comment (LeSeulArtichaut) [reply src/librustc_mir/transform/check_unsafety.rs]: I thought it would be better than fetching the caller features "for nothing", but that check can easily be removed, yes.

[the remark this answers is not in the document]
Comment (hanna-kruppe) [reply src/librustc_mir/transform/check_unsafety.rs]: I wouldn't worry about this prematurely. We can do a perf run to see the overall impact of all the changes in this PR (after removing this `if`). If it's neutral, there's no reason to try to optimize.

**Comment by LeSeulArtichaut** [comment]
@hanna-kruppe I put my code in `check_target_feature`, but it didn't pass [this test](). The reason seems to be that attributes on closures aren't checked in `check_target_feature`, which is maybe why the validation was in `codegen_fn_attrs` in the first place. Should I put it back there? Or should I still try to move it out?

**Comment by hanna-kruppe** [comment]
Sorry if I was unclear, I had two different refactorings in mind:

1. Leaving the validation of target features to `codegen_fn_attrs`, just change *where* `CheckAttrVisitor` calls `codegen_fn_attrs` to perform that validation (from a special one-off in `check_attributes` to `check_target_feature`). This should not affect closures IIUC, just remove the need to keep two conditions in sync. I would like that to happen in this PR.
2. Move the code doing the validation of target features. I didn't realize before that this `check_target_feature` does not look at closures and I believe that indicates a bug, but in any case, this change is larger and more risky, so no need to do it in this PR.

**Comment by LeSeulArtichaut** [comment]
`#[target_feature]` doesn't seem to be the only attribute that relies on `codegen_fn_attrs`. IIUC, there is also `#[inline]` and `#[track_caller]`. Should I call `codegen_fn_attrs` if I encounter any of those?

**Comment by hanna-kruppe** [comment]
Oh no, what a mess. I'll have to go back to the drawing board and probably talk to other folks to figure out how to (if at all) this whole affair should be refactored. Sorry for the noise, let's just leave it as it is now. But please add a comment to the `codegen_fn_attrs` call explaining why we need it and that it validates multiple attributes at once, so that the next person touching this code won't have to rediscover it.

I am somewhat worried that this will change how the other attributes are validated (e.g., that `inline` will now get validated on methods in cases where it previously wasn't) but I have not been able to showcase this, so maybe it happens to work out okay today.

**Comment by LeSeulArtichaut** [comment]
Rebased onto #71205, so this is now blocked on #71205

**Comment by LeSeulArtichaut** [comment]
CI passed 🎉

@hanna-kruppe I think this is ready for your review, and hopefully for merging

**Comment by LeSeulArtichaut** [comment]
I’ll need to squash the commits though :rolling_eyes:

File: src/librustc_middle/ty/error.rs [hunk 31]
```diff
@@ -58,6 +58,8 @@ pub enum TypeError<'tcx> {
     ConstMismatch(ExpectedFound<&'tcx ty::Const<'tcx>>),
 
     IntrinsicCast,
+    /// Safe `#[target_feature]` functions are not assignable to safe function pointers.
+    TargetFeatureCast(Box<[Span]>),
```
Comment (hanna-kruppe) [inline src/librustc_middle/ty/error.rs:62]: I suspect it would be enough to store the DefId of the function item, the error reporting code can get those spans from the DefId. Does that make sense to you or did I miss anything?

To be clear: I'd consider this an optional clean-up, I don't want to hold up the PR over it, but it would be good to leave a TODO comment so the idea doesn't get lost entirely.

File: src/test/ui/rfcs/rfc-2396-target_feature-11/check-pass.rs [hunk 32]
```diff
@@ -0,0 +1,49 @@
+// Tests the new rules added by RFC 2396, including:
+// - applying `#[target_feature]` to safe functions is allowed
+// - calling functions with `#[target_feature]` is allowed in
+//   functions which have (at least) the same features
+// - calling functions with `#[target_feature]` is allowed in
+//   unsafe contexts
+// - functions with `#[target_feature]` can coerce to unsafe fn pointers
+
+// check-pass
```
Comment (hanna-kruppe) [inline src/test/ui/rfcs/rfc-2396-target_feature-11/check-pass.rs:9]: PR checks only test on x86(_64) builders, right? I think this file needs an `only-x86_64` annotation to avoid errors about unknown target features on other builders.

The same probably applies to all other tests.

File: src/test/ui/target-feature/invalid-attribute.rs [hunk 33]
```diff
@@ -24,12 +24,6 @@
 //~^ ERROR malformed `target_feature` attribute
 unsafe fn foo() {}
 
-#[target_feature(enable = "sse2")]
-//~^ ERROR `#[target_feature(..)]` can only be applied to `unsafe` functions
-//~| NOTE can only be applied to `unsafe` functions
-fn bar() {}
-//~^ NOTE not an `unsafe` function
-
```
Comment (hanna-kruppe) [inline src/test/ui/target-feature/invalid-attribute.rs:32]: Why was this removed? It overlaps with the new feature-gate test now, but (at least in the current state of the PR) I see no reason to remove this, it's just unnecessary churn. Let's just leave it be, it doesn't do any harm.

**Review by hanna-kruppe — CHANGES_REQUESTED** [review]
Yes, squashing is needed, and I also noticed a few more small things after reading this with a fresh eye. Apart from that, it looks ready. Thanks for working on this and pushing it through so many rounds of review and bug fixes, we're almost there :)

File: src/test/ui/target-feature/invalid-attribute.rs (hunk 33 above)
Comment (LeSeulArtichaut) [reply src/test/ui/target-feature/invalid-attribute.rs:32]: IIRC Centril asked me to remove it. I think both letting it and removing it makes sense, but I think I shouldn't remove it if the other test that exercises it only gets executed on certain platforms.

File: src/librustc_middle/ty/error.rs [hunk 34]
```diff
@@ -193,7 +199,8 @@ impl<'tcx> TypeError<'tcx> {
         use self::TypeError::*;
         match self {
             CyclicTy(_) | UnsafetyMismatch(_) | Mismatch | AbiMismatch(_) | FixedArraySize(_)
-            | Sorts(_) | IntMismatch(_) | FloatMismatch(_) | VariadicMismatch(_) => false,
+            | Sorts(_) | IntMismatch(_) | FloatMismatch(_) | VariadicMismatch(_)
+            | TargetFeatureCast(_) => false,
```
[the remark this answers is not in the document]
Comment (LeSeulArtichaut) [reply src/librustc_middle/ty/error.rs:203]: Well, there is no particular reason. This is my first ~~major~~ non-minor contribution to the compiler, so I basically searched for things similar to what I tried to implement. I wonder if we could merge the two though, since I use the distinction between the two to have an adapted error message.

File: src/test/ui/target-feature/invalid-attribute.rs (hunk 33 above)
Comment (hanna-kruppe) [reply src/test/ui/target-feature/invalid-attribute.rs:32]: This test only gets executed on x86 targets, too (see all the ignore-* lines at the top). There's just no target-independent way to test these attributes, since there's no target feature names that are recognized on all platforms.

**Comment by LeSeulArtichaut** [comment]
I didn't want to re-build `rustc` completely (it takes at least an hour on my machine), so I did some pretty wacky stuff with the tests. Hopefully it still passes, but I wouldn't be surprised if it doesn't 😅

> Yes, squashing is needed

I tried to divide the PR in 4 commits to still have something a bit more tidy than a single commit, but I don't mind squashing again if it is deemed necessary.

> Thanks for working on this and pushing it through so many rounds of review and bug fixes

Eh, thanks to you for actually doing so many rounds of review! Also thanks to @Centril, even if he isn't here anymore :/

**Comment by petrochenkov** [comment]
Do "target-feature" functions implement the `Fn()` family of traits?

If they do, then it looks like unsoundness because the unsafety check is performed before monomorphization.

If they don't, then it means that target-feature functions are still unsafe from the type system point of view, it's just hidden syntactically (and special-cased in the unsafety checker), so the special coercion restriction is not necessary.

**Comment by hanna-kruppe** [comment]
I think you're right, this PR doesn't change that safe function items implement `Fn*` traits and this should lead to unsoundness. However, this is a concern about the contents of the RFC and belongs on the tracking issue.

I'd prefer to merge this PR regardless, the changes implemented here will probably be unaffected by whatever we end up doing to solve this soundness problem, and having 90% of an implementation in-tree is better than letting it bit-rot until we figure out the remaining 10%.
